diff --git a/node_modules/@codesandbox/sandpack-react/dist/esm/components/CodeEditor/CodeMirror.js b/node_modules/@codesandbox/sandpack-react/dist/esm/components/CodeEditor/CodeMirror.js
index 7789588..74e7011 100644
--- a/node_modules/@codesandbox/sandpack-react/dist/esm/components/CodeEditor/CodeMirror.js
+++ b/node_modules/@codesandbox/sandpack-react/dist/esm/components/CodeEditor/CodeMirror.js
@@ -14,7 +14,7 @@ import { getFileName } from "../../utils/stringUtils";
 import { highlightDecorators } from "./highlightDecorators";
 import { highlightInlineError } from "./highlightInlineError";
 import { getCodeMirrorLanguage, getEditorTheme, getSyntaxHighlight, useCombinedRefs, } from "./utils";
-export const CodeMirror = React.forwardRef(({ code, filePath, fileType, onCodeUpdate, showLineNumbers = false, showInlineErrors = false, wrapContent = false, editorState = "pristine", readOnly = false, decorators, }, ref) => {
+export const CodeMirror = React.forwardRef(({ code, filePath, fileType, onCodeUpdate, onCodeSave, showLineNumbers = false, showInlineErrors = false, wrapContent = false, editorState = "pristine", readOnly = false, decorators, }, ref) => {
     const wrapper = React.useRef(null);
     const cmView = React.useRef();
     const { theme, themeId } = useSandpackTheme();
@@ -37,6 +37,15 @@ export const CodeMirror = React.forwardRef(({ code, filePath, fileType, onCodeUp
                 key: "Shift-Tab",
                 run: indentLess,
             },
+            {
+                key: "Ctrl-S",
+                run: (instance) => {
+                    if (instance && onCodeSave) {
+                        onCodeSave(instance.state.doc.sliceString(0, instance.state.doc.length));
+                    }
+                    return true;
+                },
+            },
             {
                 key: "Escape",
                 run: () => {
diff --git a/node_modules/@codesandbox/sandpack-react/dist/esm/components/Navigator/index.js b/node_modules/@codesandbox/sandpack-react/dist/esm/components/Navigator/index.js
index 9d167b3..3baed7b 100644
--- a/node_modules/@codesandbox/sandpack-react/dist/esm/components/Navigator/index.js
+++ b/node_modules/@codesandbox/sandpack-react/dist/esm/components/Navigator/index.js
@@ -3,7 +3,7 @@ import * as React from "react";
 import { useSandpack } from "../../hooks/useSandpack";
 import { BackwardIcon, ForwardIcon, RefreshIcon } from "../../icons";
 import { splitUrl } from "./utils";
-export const Navigator = ({ clientId, onURLChange, }) => {
+export const Navigator = ({ clientId, onURLChange, appendMenu: AppendMenu, }) => {
     var _a;
     const [baseUrl, setBaseUrl] = React.useState("");
     const { sandpack, dispatch, listen } = useSandpack();
@@ -49,13 +49,17 @@ export const Navigator = ({ clientId, onURLChange, }) => {
     const handleForward = () => {
         dispatch({ type: "urlforward" });
     };
-    return (React.createElement("div", { className: c("navigator") },
-        React.createElement("button", { "aria-label": "Go back one page", className: c("button", "icon"), disabled: !backEnabled, onClick: handleBack, type: "button" },
-            React.createElement(BackwardIcon, null)),
-        React.createElement("button", { "aria-label": "Go forward one page", className: c("button", "icon"), disabled: !forwardEnabled, onClick: handleForward, type: "button" },
-            React.createElement(ForwardIcon, null)),
-        React.createElement("button", { "aria-label": "Refresh page", className: c("button", "icon"), onClick: handleRefresh, type: "button" },
-            React.createElement(RefreshIcon, null)),
-        React.createElement("input", { "aria-label": "Current Sandpack URL", className: c("input"), name: "Current Sandpack URL", onChange: handleInputChange, onKeyDown: handleKeyDown, type: "text", value: relativeUrl })));
+    return (
+        React.createElement("div", { className: c("navigator") },
+            React.createElement("button", { "aria-label": "Go back one page", className: c("button", "icon"), disabled: !backEnabled, onClick: handleBack, type: "button" },
+                React.createElement(BackwardIcon, null)),
+            React.createElement("button", { "aria-label": "Go forward one page", className: c("button", "icon"), disabled: !forwardEnabled, onClick: handleForward, type: "button" },
+                React.createElement(ForwardIcon, null)),
+            React.createElement("button", { "aria-label": "Refresh page", className: c("button", "icon"), onClick: handleRefresh, type: "button" },
+                React.createElement(RefreshIcon, null)),
+            React.createElement("input", { "aria-label": "Current Sandpack URL", className: c("input"), name: "Current Sandpack URL", onChange: handleInputChange, onKeyDown: handleKeyDown, type: "text", value: relativeUrl }),
+            AppendMenu && React.createElement(AppendMenu, null)
+        )
+    );
 };
 //# sourceMappingURL=index.js.map
\ No newline at end of file
diff --git a/node_modules/@codesandbox/sandpack-react/dist/esm/components/Preview/index.js b/node_modules/@codesandbox/sandpack-react/dist/esm/components/Preview/index.js
index 9b58485..c2a3979 100644
--- a/node_modules/@codesandbox/sandpack-react/dist/esm/components/Preview/index.js
+++ b/node_modules/@codesandbox/sandpack-react/dist/esm/components/Preview/index.js
@@ -9,7 +9,7 @@ import { generateRandomId } from "../../utils/stringUtils";
 import { Navigator } from "../Navigator";
 import { RefreshButton } from "./RefreshButton";
 export { RefreshButton };
-export const SandpackPreview = ({ customStyle, showNavigator = false, showRefreshButton = true, showOpenInCodeSandbox = true, showSandpackErrorOverlay = true, viewportSize = "auto", viewportOrientation = "portrait", }) => {
+export const SandpackPreview = ({ customStyle, navigatorComponent: NavigatorComponent, showNavigator = false, showRefreshButton = true, showOpenInCodeSandbox = true, showSandpackErrorOverlay = true, viewportSize = "auto", viewportOrientation = "portrait", }) => {
     const { sandpack, listen } = useSandpack();
     const [iframeComputedHeight, setComputedAutoHeight] = React.useState(null);
     const { status, registerBundler, unregisterBundler, errorScreenRegisteredRef, openInCSBRegisteredRef, loadingScreenRegisteredRef, } = sandpack;
@@ -45,7 +45,7 @@ export const SandpackPreview = ({ customStyle, showNavigator = false, showRefres
             ...viewportStyle,
             display: status !== "idle" ? "flex" : "none",
         } },
-        showNavigator ? (React.createElement(Navigator, { clientId: clientId.current, onURLChange: handleNewURL })) : null,
+        showNavigator ? (React.createElement(NavigatorComponent || Navigator, { clientId: clientId.current, onURLChange: handleNewURL })) : null,
         React.createElement("div", { className: c("preview-container") },
             React.createElement("iframe", { ref: iframeRef, className: c("preview-iframe"), style: {
                     // set height based on the content only in auto mode
diff --git a/node_modules/@codesandbox/sandpack-react/dist/esm/contexts/sandpackContext.js b/node_modules/@codesandbox/sandpack-react/dist/esm/contexts/sandpackContext.js
index 7d3a0d1..5a88e07 100644
--- a/node_modules/@codesandbox/sandpack-react/dist/esm/contexts/sandpackContext.js
+++ b/node_modules/@codesandbox/sandpack-react/dist/esm/contexts/sandpackContext.js
@@ -193,6 +193,9 @@ class SandpackProvider extends React.PureComponent {
             });
             this.updateClients();
         };
+        this.updateOpenPaths = (paths) => {
+            this.setState({ openPaths: paths });
+        };
         this.dispatchMessage = (message, clientId) => {
             if (this.state.sandpackStatus !== "running") {
                 console.warn("dispatch cannot be called while in idle mode");
@@ -223,16 +226,26 @@ class SandpackProvider extends React.PureComponent {
                         this.unsubscribeQueuedListeners[clientId] || {};
                     this.queuedListeners[clientId][listenerId] = listener;
                     const unsubscribeListener = () => {
-                        if (this.queuedListeners[clientId][listenerId]) {
+                        if (this.queuedListeners[clientId] && this.queuedListeners[clientId][listenerId]) {
                             // unsubscribe was called before the client was instantiated
                             // common example - a component with autorun=false that unmounted
-                            delete this.queuedListeners[clientId][listenerId];
+                            // delete this.queuedListeners[clientId][listenerId];
+                            const { [listenerId]: removed, ...rest } = this.queuedListeners[clientId];
+                            this.queuedListeners = {
+                                ...this.qqueuedListeners,
+                                [clientId]: rest
+                            }
                         }
-                        else if (this.unsubscribeQueuedListeners[clientId][listenerId]) {
+                        else if (this.unsubscribeQueuedListeners[clientId] && this.unsubscribeQueuedListeners[clientId][listenerId]) {
                             // unsubscribe was called for a listener that got added before the client was instantiated
                             // call the unsubscribe function and remove it from memory
                             this.unsubscribeQueuedListeners[clientId][listenerId]();
-                            delete this.unsubscribeQueuedListeners[clientId][listenerId];
+                            // delete this.unsubscribeQueuedListeners[clientId][listenerId];
+                            const { [listenerId]: removed, ...rest } = this.unsubscribeQueuedListeners[clientId];
+                            this.unsubscribeQueuedListeners = {
+                                ...this.unsubscribeQueuedListeners,
+                                [clientId]: rest
+                            }
                         }
                     };
                     return unsubscribeListener;
@@ -269,8 +282,9 @@ class SandpackProvider extends React.PureComponent {
             this.setState({ files }, this.updateClients);
         };
         this._getSandpackState = () => {
-            const { files, activePath, openPaths, startRoute, bundlerState, editorState, error, sandpackStatus, environment, } = this.state;
+            const { files, activePath, openPaths, startRoute, bundlerState, editorState, error, sandpackStatus, environment, origin, } = this.state;
             return {
+                origin,
                 files,
                 environment,
                 openPaths,
@@ -297,10 +311,12 @@ class SandpackProvider extends React.PureComponent {
                 unregisterBundler: this.unregisterBundler,
                 updateCurrentFile: this.updateCurrentFile,
                 updateFile: this.updateFile,
+                updateOpenPaths: this.updateOpenPaths,
             };
         };
         const { activePath, openPaths, files, environment, } = getSandpackStateFromProps(props);
         this.state = {
+            origin: files,
             files,
             environment,
             openPaths,
@@ -368,7 +384,21 @@ class SandpackProvider extends React.PureComponent {
                 JSON.stringify(this.props.customSetup)) {
             const { activePath, openPaths, files, environment, } = getSandpackStateFromProps(this.props);
             /* eslint-disable react/no-did-update-set-state */
-            this.setState({ activePath, openPaths, files, environment });
+            // this.setState({ activePath, openPaths, files, environment });
+            this.setState({
+                activePath,
+                openPaths,
+                environment,
+                origin: files,
+                files: {
+                    ...files,
+                    ...Object.fromEntries(
+                        Object.entries(this.state.files).filter(([key, _value]) => {
+                            return files.hasOwnProperty(key) && this.state.files[key].code !== this.state.origin[key].code;
+                        })
+                    ),
+                },
+            });
             if (this.state.sandpackStatus !== "running") {
                 return;
             }
